### PR TITLE
gh-122234: fix accuracy issues for sum()

### PR DESCRIPTION
**Pull request by skirpichev**
* Use compensated summation for complex sums with floating-point items. This amends #121176.

* sum() specializations for floats and complexes now use PyLong_AsDouble() instead of PyLong_AsLongAndOverflow() and compensated summation as well.

* Issue: gh-122234

### CODE REVIEW

**Comment by picnixz** [comment]
I'm asking Victor and Raymond since I know that Victor has been reviewing Sergey's PR (and I think you also merged it) and Raymond since Sergey asked him a question on the issue. Are there other people I should ask?

File: Python/bltinmodule.c [hunk 1]
```diff
@@ -2687,11 +2687,9 @@ builtin_sum_impl(PyObject *module, PyObject *iterable, PyObject *start)
...
-                int overflow;
-                value = PyLong_AsLongAndOverflow(item, &overflow);
-                if (!overflow) {
-                    re_sum.hi += (double)value;
+                double value = PyLong_AsDouble(item);
+                if (value != -1.0 || !PyErr_Occurred()) {
+                    re_sum = cs_add(re_sum, value);
                     Py_DECREF(item);
                     continue;
                 }
```
Comment (ericsnowcurrently) [inline Python/bltinmodule.c:2695]: We should be sure the exception gets handled:

```suggestion
                double value = PyLong_AsDouble(item);
                if (value != -1.0 || !PyErr_Occurred()) {
                    re_sum = cs_add(re_sum, value);
                    Py_DECREF(item);
                    continue;
                }
                if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
                    PyErr_Clear();
                }
                else {
                    return NULL;
                }
```

File: Python/bltinmodule.c [hunk 2]
```diff
@@ -2736,19 +2734,17 @@ builtin_sum_impl(PyObject *module, PyObject *iterable, PyObject *start)
                 continue;
             }
             if (PyLong_Check(item)) {
-                long value;
-                int overflow;
-                value = PyLong_AsLongAndOverflow(item, &overflow);
-                if (!overflow) {
-                    re_sum.hi += (double)value;
+                double value = PyLong_AsDouble(item);
+                if (value != -1.0 || !PyErr_Occurred()) {
```
Comment (ericsnowcurrently) [inline Python/bltinmodule.c:2741]: Likewise, the exception should be properly handled.

File: Python/bltinmodule.c (hunk 1 above)
Comment (ericsnowcurrently) [reply Python/bltinmodule.c:2695]: Also, it would be good to add a test for this case: `item` overflows/underflows a `double`.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]

File: Python/bltinmodule.c (hunk 1 above)
Comment (skirpichev) [reply Python/bltinmodule.c:2695]: Thanks!  I think we should add a quick exit (return NULL) as an alternative.

But I doubt we should use PyErr_ExceptionMatches.  Rather, we should exit unconditionally: [OverflowError](https://docs.python.org/3/library/exceptions.html#OverflowError) is *only* possible exception at this point (according to [docs](https://docs.python.org/3.12/c-api/long.html#c.PyLong_AsDouble)).  And when we fallback to PyNumber_Add() - this exception will be raised again:
```pycon
Python 3.12.0+ (heads/3.12:2162512d71, Jan 13 2024, 11:25:20) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> sum([1.0, 10**1000])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: int too large to convert to float
```

**Comment by skirpichev** [comment]
I have made the requested changes; please review again

**Review by ericsnowcurrently — APPROVED** [review]
Seems okay to me.  However, you should have an expert on numeric stuff also take a look before this gets merged.

**Review by rhettinger — APPROVED** [review]

**Comment by vstinner** [comment]
Merged, thank you.

**Pull request merged**